SignalServiceKit/src/Messages/OWSReadReceiptManager.m

```diff
@@ -122,51 +145,246 @@ - (void)process
+
+            OWSReadReceipt *newReadReceipt =
+                [[OWSReadReceipt alloc] initWithSenderId:messageAuthorId timestamp:message.timestamp];
+
+            OWSReadReceipt *_Nullable oldReadReceipt = self.toLinkedDevicesReadReceiptMap[threadUniqueId];
+            if (oldReadReceipt && oldReadReceipt.timestamp > newReadReceipt.timestamp) {
+                // If there's an existing read receipt for the same thread with
+                // a newer timestamp, discard the new read receipt.
+                DDLogVerbose(@"%@ Ignoring redundant read receipt for linked devices.", self.tag);
+            } else {
+                DDLogVerbose(@"%@ Enqueuing read receipt for linked devices.", self.tag);
+                self.toLinkedDevicesReadReceiptMap[threadUniqueId] = newReadReceipt;
+            }
+
+            if ([self areReadReceiptsEnabled]) {
+                DDLogVerbose(@"%@ Enqueuing read receipt for sender.", self.tag);
+                NSMutableArray<NSNumber *> *_Nullable timestamps = self.toSenderReadReceiptMap[messageAuthorId];
+                if (!timestamps) {
+                    timestamps = [NSMutableArray new];
+                    self.toSenderReadReceiptMap[messageAuthorId] = timestamps;
+                }
+                [timestamps addObject:@(message.timestamp)];
+            }
 
-        NSString *messageAuthorId = message.messageAuthorId;
-        OWSAssert(messageAuthorId.length > 0);
+            [self scheduleProcessing];
+        }
+    });
+}
+
+#pragma mark - Read Receipts From Recipient
+
+- (void)processReadReceiptsFromRecipient:(OWSSignalServiceProtosReceiptMessage *)receiptMessage
+                                envelope:(OWSSignalServiceProtosEnvelope *)envelope
+{
+    OWSAssert(receiptMessage);
+    OWSAssert(envelope);
+    OWSAssert(receiptMessage.type == OWSSignalServiceProtosReceiptMessageTypeRead);
+
+    if (![self areReadReceiptsEnabled]) {
+        DDLogInfo(@"%@ Ignoring incoming receipt message as read receipts are disabled.", self.tag);
+        return;
+    }
+
+    NSString *recipientId = envelope.source;
+    OWSAssert(recipientId.length > 0);
+
+    PBArray *timestamps = receiptMessage.timestamp;
+
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        [self.dbConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
+            for (int i = 0; i < timestamps.count; i++) {
+                UInt64 timestamp = [timestamps uint64AtIndex:i];
+
+                NSArray<TSOutgoingMessage *> *messages
+                    = (NSArray<TSOutgoingMessage *> *)[TSInteraction interactionsWithTimestamp:timestamp
+                                                                                       ofClass:[TSOutgoingMessage class]
+                                                                               withTransaction:transaction];
+                OWSAssert(messages.count <= 1);
```

Comment:
Interesting.  I'm not going to address this in this PR, since this isn't a regression (nor was it changed in the "message processing edge cases" PR).  

> Maybe we should always record the raw envelope timestamp, but include an explicit "sorting offset" property on interactions, which is used as a "tie breaker".

Tempting, but that would require rebuilding a expensive database view.

> Or, you know, we probably want to implement an attachment+caption collection cell eventually anyway...

Post-JSQ, right?

For now, I'm going to modify the "delivery receipt" logic to no longer assume that there's only one interaction for a given timestamp, which always problematic.